design/src/rfcs/rfc0022_refine_builder.md

```diff
@@ -0,0 +1,556 @@
+
+```rust
+fn partial_setup<Op1, Op2, Op3, Op4, Op5, Op6>(
+    builder: PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6>,
+) -> PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6> {
+    builder.get_server_statistics(get_server_statistics)
+}
+```
+
+That looks innocent, but it doesn't fly:
+
+```text
+error[E0308]: mismatched types
+  --> pokemon-service/src/bin/pokemon-service.rs:31:5
+   |
+28 | fn partial_setup<Op1, Op2, Op3, Op4, Op5, Op6>(
+   |                                 --- this type parameter
+29 |     builder: PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6>,
+30 | ) -> PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6> {
+   |      --------------------------------------------------- expected `PokemonServiceBuilder<Op1, Op2, Op3, Op4, Op5, Op6>` because of return type
+31 |     builder.get_server_statistics(get_server_statistics)
+   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected type parameter `Op4`, found struct `Operation`
+   |
+   = note: expected struct `PokemonServiceBuilder<_, _, _, Op4, _, _, _>`
+              found struct `PokemonServiceBuilder<_, _, _, Operation<IntoService<GetServerStatistics, fn(GetServerStatisticsInput, Extension<Arc<State>>) -> impl Future<Output = GetServerStatisticsOutput> {get_server_statistics}>>, _, _, _>
+```
+
+By registering a handler we have changed the corresponding `OpX` generic parameter.  
+Fixing this error requires some non-trivial type gymnastic - I gave up after trying for ~15 minutes. 
+
+### Cut them down: going from 2N+1 to 2 generic parameters
+
+The previous two examples should have convinced you that the 2N+1 generic parameters on `PokemonServiceBuilder` harm the ergonomics of our API.  
+Can we get rid of them?
+
+Yes! Let's look at one possible approach:
+
+```rust
+pub struct PokemonServiceBuilder<Body, Plugin> {
+    check_health: Option<Route<Body>>,
+    do_nothing: Option<Route<Body>>,
+    get_pokemon_species: Option<Route<Body>>,
+    get_server_statistics: Option<Route<Body>>,
+    capture_pokemon: Option<Route<Body>>,
+    get_storage: Option<Route<Body>>,
+    plugin: Plugin,
+}
+```
+
+We no longer store the raw handlers inside `PokemonServiceBuilder`.
+We eagerly upgrade the operation handlers to a `Route` instance when they are registered with the builder. The existing API performs the upgrade when `build` is called, forcing `PokemonServiceBuilder` to store the raw handlers and keep two generic parameters around (`OpX` and `ExtsX`) for each operation.  
+
+There is one downside to this alternative approach - all plugins must be specified upfront, when creating an instance of the builder.
+Plugins cannot be modified through `PokemonServiceBuilder`'s API if we want to guarantee that all operation handlers are upgraded to a `Route` using the same plugins.
+
+```rust
+impl PokemonService<()> {
+    /// Constructs a builder for [`PokemonService`].
+    pub fn builder<Body, Plugin>(plugin: Plugin) -> PokemonServiceBuilder<Body, Plugin> {
```

Comment:
We could also do a two-phase builder which keeps the user in the method chain, something like:

```rust
let app = PokemonService::builder() // : PokemonServicePluginBuilder<IdentityPlugin>
    .print()
    /* ... */
    .finalize_plugins() // PokemonServiceBuilder<B, PluginStack<IdentityPlugin, ..., PrintPlugin>>
    .get_pokemon_species(/* handler */)
    ...
    .build();
```

Reply:
Yeah, this could work as well, with a `PokemonService::builder_without_plugins()` for the case where the developer doesn't want to register any plugin and `PokemonService::builder().finalize_plugins()` would be ugly.